ui/gtk4.c

```diff
@@ -0,0 +1,2194 @@
+        if (!s->ptr_owner->window) {
+            g_simple_action_set_state(s->grab_action,
+                                      g_variant_new_boolean(FALSE));
+        } else {
+            gd_ungrab_pointer(s);
+        }
+    }
+    for (i = 0; i < s->nb_vcs; i++) {
+        VirtualConsole *vc = &s->vc[i];
+        gd_update_cursor(vc);
+    }
+}
+
+/** GTK Events **/
+
+static gboolean gd_window_close(GtkWidget *widget, GtkDisplayState *s)
+{
+    bool allow_close = true;
+
+    if (s->opts->has_window_close && !s->opts->window_close) {
+        allow_close = false;
+    }
+
+    if (allow_close) {
+        qmp_quit(NULL);
+        g_application_quit(G_APPLICATION(s->app));
+    }
+
+    return TRUE;
+}
+
+static void gd_set_ui_refresh_rate(VirtualConsole *vc, int refresh_rate)
+{
+    QemuUIInfo info;
+
+    info = *dpy_get_ui_info(vc->gfx.dcl.con);
+    info.refresh_rate = refresh_rate;
+    dpy_set_ui_info(vc->gfx.dcl.con, &info, true);
+}
+
+static void gd_set_ui_size(VirtualConsole *vc, gint width, gint height)
+{
+    QemuUIInfo info;
+
+    info = *dpy_get_ui_info(vc->gfx.dcl.con);
+    info.width = width;
+    info.height = height;
+    dpy_set_ui_info(vc->gfx.dcl.con, &info, true);
+}
+
+static gboolean gd_render_event(GtkGLArea *area, GdkGLContext *context,
+                                VirtualConsole *vc)
+{
+    if (vc->gfx.gls) {
+        gd_gl_area_draw(vc);
+    }
+    return TRUE;
+}
+
+static void gd_resize_event(GtkGLArea *area,
+                            gint width, gint height, VirtualConsole *vc)
+{
+    gd_set_ui_size(vc, width, height);
+}
+
+void gd_update_monitor_refresh_rate(VirtualConsole *vc, GtkWidget *widget)
+{
+    GtkNative *native = gtk_widget_get_native(widget);
+    GdkSurface *surface = gtk_native_get_surface(native);
+    int refresh_rate;
+
+    if (surface) {
+        GdkDisplay *dpy = gtk_widget_get_display(widget);
+        GdkMonitor *monitor = gdk_display_get_monitor_at_surface(dpy, surface);
+        refresh_rate = gdk_monitor_get_refresh_rate(monitor); /* [mHz] */
+    } else {
+        refresh_rate = 0;
+    }
+
+    gd_set_ui_refresh_rate(vc, refresh_rate);
+
+    /* T = 1 / f = 1 [s*Hz] / f = 1000*1000 [ms*mHz] / f */
+    vc->gfx.dcl.update_interval = refresh_rate ?
+        MIN(1000 * 1000 / refresh_rate, GUI_REFRESH_INTERVAL_DEFAULT) :
+        GUI_REFRESH_INTERVAL_DEFAULT;
+}
+
+static void gd_draw_event(GtkDrawingArea *widget, cairo_t *cr, int ww, int wh,
+                          void *opaque)
+{
+    VirtualConsole *vc = opaque;
+    GtkDisplayState *s = vc->s;
+    int mx, my;
+    int fbw, fbh;
+
+    if (vc->gfx.gls) {
+        if (gtk_use_gl_area) {
+            /* invoke render callback please */
+            return;
+        } else {
+            abort();
+        }
+    }
+
+    if (!gtk_widget_get_realized(GTK_WIDGET(widget))) {
+        return;
+    }
+    if (!vc->gfx.ds) {
+        return;
+    }
+    if (!vc->gfx.surface) {
+        return;
```

Comment:
This function also have early returns, any of these may be wort some kind of warning message?